README.md
# 整合注單模組

整合注單模組主要處理三個工作

+ 原生注單抓取 (fetcher)： 將第三方遊戲的原生注單抓回至本機儲存
+ 原生注單轉換 (converter)： 轉換抓回來原生注單為整合式注單
+ 整合注單查詢： 為整合式注單提供方便的查詢功能

## 安裝
因為是私有版控庫，安裝此 package 的專案必需在自己的 composer.json 先定義版控庫來源

    "repositories": [
        {
            "type": "git",
            "url": "git@git.sp168.cc:super-platform/united-ticket.git"
        },
    ],

接著就可以透過下列指令進行安裝

    composer require super-platform/united-ticket

如果 Laravel 版本在 5.4 以下，你必需手動追加 ServerProvider

    // config/app.php
    'providers' => [
        ...
        SuperPlatform\UnitedTicket\UnitedTicketServiceProvider::class,
    ],

如果不確定，就老實的使用手動追加最保險的方式

然後必須手動下 `php artisan vendor:publish` 選擇此專案進行發佈載入 migration

## 使用方法

### 原生注單同步(抓取 & 轉換)

#### 可用的 artisan 指令

    // 列出目前可以進行同步的遊戲站識別碼
    php artisan united-ticket:stations 

    // 同步指定{遊戲站}-{帳號}-{指定開始時間}-{指定結束時間} 的原生注單
    php artisan ticket-integrator:transfer {station} {username} {--startTime} {--endTime}

#### 使用排程定時同步

    // app/Console/Kernel.php

    use SuperPlatform\StationWallet\StationLoginRecord;
    use SuperPlatform\UnitedTicket\Jobs\RawTicketSyncJob;

    protected function schedule(Schedule $schedule)
    { 
        /* 注單抓取範圍 */
        $all_bet_range_days = 14; 
        $sa_gaming_range_days = 7;
        
        // 各遊戲站當日有登入過的會員
        $users = StationLoginRecord::getRecord(['status' => 'clicked'])->groupBy('station');
        
        $schedule->call(function () use ($users) {
            // 同步歐博的帳號注單
            $usersInAllBet = array_get($users, 'all_bet', []);
            $usersInGame->each(function ($item) {
                    $userInGame = [
                        'username' => $item->account,
                        'user_id' => $item->user_id
                    ];
        
                    dispatch(new RawTicketSyncJob('all_bet', $userInGame, now()->subDay(14)->toDateTimeString(), now()->toDateTimeString()))
                        ->onQueue('united-ticket');
                });
                
            // 同步沙龍的帳號注單
            $usersInSaGaming = array_get($users, 'sa_gaming', []);
            $usersInGame->each(function ($item) {
                    $userInGame = [
                        'username' => $item->account,
                        'user_id' => $item->user_id
                    ];
        
                    dispatch(new RawTicketSyncJob('sa_gaming', $userInGame, now()->subDay(7)->toDateTimeString(), now()->toDateTimeString()))
                        ->onQueue('united-ticket');
                });
            
            
        })
        // 每 1 分鐘執行這個任務
        ->everyMinute();
    }
    
#### 使用手動同步

    use SuperPlatform\UnitedTicket\Jobs\RawTicketSyncJob;
    
    dispatch(new RawTicketSyncJob($station, $username, $startTime, $endTime)
                    ->onQueue('united-ticket'));
    
### 整合注單查詢

#### 可用的 artisan 指令

    // 建立 {遊戲站} {--指定時間點} 此時間點的「上一個刻鐘區段」的預查結果
    php artisan united-ticket:stored-block {--station} {--start} {--end}
    
#### 使用排程定時建立預查結果(不指定遊戲站就會選全部)

    // app/Console/Kernel.php

    use SuperPlatform\UnitedTicket\Jobs\StoredBlockJob;

    protected function schedule(Schedule $schedule)
    {
        $schedule->call(function () use ($stations) {
            dispatch((new StoredBlockJob())->init())->onQueue('united-ticket-block');
        })
        // 每 15 分鐘執行這個任務
        ->everyFifteenMinutes()
        // 在執行之前戳一下
        ->pingBefore($url)
        // 任務完成後戳一下
        ->thenPing($url);
    }
        
#### 查詢範例

查詢遊戲站 sa_gaming，帳號 hero 在下列時段

+ `2017-08-01 12:00:00 ~ 2017-08-01 12:15:00`

的贏輸結果
    
出來結果範例

| user_identify              | station   | game_type | bet_type | sum_raw_bet | sum_valid_bet | sum_rolling | sum_winnings | time_span_begin     | time_span_end       | depth_1_identify |depth_1_ratio|
|----------------------------|-----------|-----------|----------|-------------|---------------|-------------|--------------|---------------------|---------------------|------------------|-------------|
| 000zmep5qb9asdktz10f60xmm2 | sa_gaming | sicbo     | big      | 340         | 340           | 340         | 680          | 2017-08-01 12:13:14 | 2017-08-02 22:23:23 | root_id          |0.97         |   
| 000zmep5qb9asdktz10f60xmm2 | sa_gaming | sicbo     | big      | 520         | 520           | 520         | 1040         | 2017-08-02 22:23:24 | 2017-08-03 04:44:32 | user_1_id        |0.98         | 
| 000zmep5qb9asdktz10f60xmm2 | sa_gaming | slot      | spin     | 246         | 246           | 246         | 442          | 2017-08-01 12:13:14 | 2017-08-02 22:23:23 | user_2_id        |0.99         |
| 000zmep5qb9asdktz10f60xmm2 | sa_gaming | slot      | spin     | 333         | 333           | 333         | -333         | 2017-08-02 22:23:24 | 2017-08-03 04:44:32 | user_3_id        |0.98         |
 
 ### 整合注單更新
 
 ### 可用的 artisan 指令
    // 每日更新today_tickets和yesterday_tickets redis資料庫資料
    php artisan united-ticket:daily-refresh
    
    // 每週更新this_week_tickets和last_week_tickets redis資料庫資料
    php artisan united-ticket:weekly-refresh
    
    // 每月更新this_month_tickets和last_month_tickets redis資料庫資料
    php artisan united-ticket:monthly-refresh
    
#### 使用排程定時更新資料

    // app/Console/Kernel.php

    use SuperPlatform\UnitedTicket\Jobs\DailyRefreshBlockJob;
    use SuperPlatform\UnitedTicket\Jobs\WeeklyRefreshBlockJob;
    use SuperPlatform\UnitedTicket\Jobs\MonthlyRefreshBlockJob;

    protected function schedule(Schedule $schedule)
    {
        $schedule->call(function () {
            dispatch(new DailyRefreshBlockJob())->onQueue('daily-refresh-block'));
        })
        // 每日00:00做更新
        ->dailyAt('00:00')
        // 在執行之前戳一下
        ->pingBefore($url)
        // 任務完成後戳一下
        ->thenPing($url);
        
        $schedule->call(function () {
            dispatch(new WeeklyRefreshBlockJob())->onQueue('weekly-refresh-block'));
        })
        // 每週一 00:00做更新
        ->weeklyOn(0, '00:00')
        // 在執行之前戳一下
        ->pingBefore($url)
        // 任務完成後戳一下
        ->thenPing($url);
        
        $schedule->call(function () {
            dispatch(new MonthlyRefreshBlockJob())->onQueue('monthly-refresh-block'));
        })
        // 每月1號 00:00做更新
        ->monthlyOn(1, '00:00')
        // 在執行之前戳一下
        ->pingBefore($url)
        // 任務完成後戳一下
        ->thenPing($url);
    }


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.25.5"></a>
## [1.25.5](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.4...v1.25.5) (2020-06-24)


### Bug Fixes

* 越南彩新增幣值轉換( [@pg](https://git.sp168.cc/pg) ) ([74d3205](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/74d3205))



<a name="1.25.4"></a>
## [1.25.4](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.2...v1.25.4) (2020-06-24)


### Bug Fixes

* 將 br-push 也補上版本號碼 ([b6ad575](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/b6ad575))
* 越南彩新增幣值轉換( [@pg](https://git.sp168.cc/pg) ) ([99bd0dd](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/99bd0dd))



<a name="1.25.1"></a>
## [1.25.1](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.0...v1.25.1) (2020-06-19)



<a name="1.25.3"></a>
## [1.25.3](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.2...v1.25.3) (2020-06-24)


### Bug Fixes

* 將 br-push 也補上版本號碼 ([b6ad575](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/b6ad575))
* 越南彩新增幣值轉換( [@pg](https://git.sp168.cc/pg) ) ([99bd0dd](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/99bd0dd))



<a name="1.25.1"></a>
## [1.25.1](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.0...v1.25.1) (2020-06-19)



<a name="1.25.2"></a>
## [1.25.2](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.0...v1.25.2) (2020-06-20)


### Bug Fixes

* 增加越南彩退水值 ([@awei](https://git.sp168.cc/awei)) ([aae6845](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/aae6845))



<a name="1.25.1"></a>
## [1.25.1](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.25.0...v1.25.1) (2020-06-20)


### Bug Fixes

* 增加越南彩退水值 ([@awei](https://git.sp168.cc/awei)) ([aae6845](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/aae6845))



<a name="1.25.0"></a>
# [1.25.0](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.61...v1.25.0) (2020-06-19)


### Features

* 新增 npm run release-branch 指令，可快速產生對應版本分支, 與一併 push tag 指令 ([9133357](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/9133357))



<a name="1.24.61"></a>
## [1.24.61](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.60...v1.24.61) (2020-06-19)


### Bug Fixes

* mg棋牌開牌結果修改 ([@kevin](https://git.sp168.cc/kevin)) ([ebbae12](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/ebbae12))



<a name="1.24.60"></a>
## [1.24.60](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.59...v1.24.60) (2020-06-10)


### Bug Fixes

* 新增DG魚蝦蟹遊戲 ([@awei](https://git.sp168.cc/awei)) ([cac2330](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/cac2330))



<a name="1.24.59"></a>
## [1.24.59](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.58...v1.24.59) (2020-06-10)


### Bug Fixes

* 修改反波膽開牌結果中的注單狀態 ([@awei](https://git.sp168.cc/awei)) ([55edeba](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/55edeba))
* 新增整合注單開賽時間欄位 ([@awei](https://git.sp168.cc/awei)) ([c821979](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/c821979))



<a name="1.24.58"></a>
## [1.24.58](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.57...v1.24.58) (2020-06-09)


### Bug Fixes

* 修改檔案名稱 ([@kevin](https://git.sp168.cc/kevin)) ([9635c4f](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/9635c4f))



<a name="1.24.57"></a>
## [1.24.57](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.56...v1.24.57) (2020-06-08)



<a name="1.24.56"></a>
## [1.24.56](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.55...v1.24.56) (2020-06-05)


### Bug Fixes

* 修正反波膽若為無效單則派彩時間判斷為取消住單時間 ([@awei](https://git.sp168.cc/awei)) ([0f0eb3c](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/0f0eb3c))



<a name="1.24.55"></a>
## [1.24.55](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.54...v1.24.55) (2020-06-04)


### Bug Fixes

* 修改注單轉換錯誤([@nick](https://git.sp168.cc/nick)) ([8f6702c](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/8f6702c))



<a name="1.24.54"></a>
## [1.24.54](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.53...v1.24.54) (2020-06-04)


### Bug Fixes

* resolve conflict([@nick](https://git.sp168.cc/nick)) ([32386c6](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/32386c6))
* 增加SF電子新遊戲 ([@awei](https://git.sp168.cc/awei)) ([7bd9418](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/7bd9418))



<a name="1.24.53"></a>
## [1.24.53](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.52...v1.24.53) (2020-06-04)


### Bug Fixes

* 修改注單轉換錯誤([@nick](https://git.sp168.cc/nick)) ([c33c047](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/c33c047))



<a name="1.24.52"></a>
## [1.24.52](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.51...v1.24.52) (2020-06-01)



<a name="1.24.51"></a>
## [1.24.51](https://git.sp168.cc/super-platform/laravel-package/united-ticket/compare/v1.24.50...v1.24.51) (2020-05-28)


### Bug Fixes

* 修改反波膽退水注單拉到foreach結束後 ([@awei](https://git.sp168.cc/awei)) ([b6b58af](https://git.sp168.cc/super-platform/laravel-package/united-ticket/commits/b6b58af))



<a name="1.24.50"></a>

src/BlockUnitedTicket.php
<?php

namespace SuperPlatform\UnitedTicket;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Redis;
use SuperPlatform\UnitedTicket\Models\BlockUnitedTicket as BUT;
use Ramsey\Uuid\Uuid;

class BlockUnitedTicket
{
    private $united_ticket_ids = [];

    public function storeBlock($station, $start = '', $end = '')
    {

        // 如果沒有指定時間區間，就以上一刻鐘為主
        $timeStart = empty($start) ? now()->subMinutes(now()->minute % 15)->subMinutes(15)->format('Y-m-d H:i:00') : $start;
        $timeEnd = empty($end) ? now()->subMinutes(now()->minute % 15)->format('Y-m-d H:i:00') : $end;

        // 從united_tickets table 取出注單並使用user_identify，station，game_scope，bet_type，depth_identify .... 做分類
        $block = $this->selectBlock($station, $timeStart, $timeEnd);

        // 使用user_identify，station，game_scope，bet_type，depth_identify組合成整合項目的id
        $block = $this->blockTransform($block, $timeStart);

        // 把tickets加到daily，weekly，monthly redis，供營運交收報表使用
//        foreach ($block as $blockTicket) {
//            Redis::command('lpush', ['today_tickets', json_encode($blockTicket)]);
//        }
//
//        foreach ($block as $blockTicket) {
//            Redis::command('lpush', ['this_week_tickets', json_encode($blockTicket)]);
//        }
//
//        foreach ($block as $blockTicket) {
//            Redis::command('lpush', ['this_month_tickets', json_encode($blockTicket)]);
//        }

        if ($block->isNotEmpty()) {
            // 把tickets存進MYSQL，保存為將來使用
            BUT::replace(
                json_decode($block->toJson(), true)
            );

//            // 修改壓縮過的注單狀態為 true
//            DB::table('united_tickets')
//                ->whereIn('id', $this->united_ticket_ids)
//                ->update(['compressed' => true]);
        }


//        Redis::command('del', ['tickets']);
    }

    /**
     * @param $station
     * @param $timeStart
     * @param $timeEnd
     * @return mixed
     */
    private function selectBlock($station, $timeStart, $timeEnd)
    {
        //---------------------------------------以下是使用REDIS，速度比較快------------------------------------------------------------------------
//        $tickets = collect(Redis::command('lrange', ['tickets', 0, -1]));
//
//        return $tickets->map(function ($ticket) {
//            return json_decode($ticket);
//        })
//        ->groupBy(['user_identify', 'station', 'game_scope',
//                'depth_1_identify', 'depth_2_identify', 'depth_3_identify',
//                'depth_4_identify', 'depth_5_identify', 'depth_6_identify',
//                'depth_7_identify', 'depth_8_identify', 'depth_9_identify',
//                'depth_10_identify', 'depth_11_identify', 'depth_12_identify'
//        ])->flatten(14) // 取到最底層的上一層
//        ->map(function ($blockTickets) use ($timeStart, $timeEnd) {
//            $sum_raw_bet = $blockTickets->sum('raw_bet');
//            $sum_valid_bet = $blockTickets->sum('valid_bet');
//            $sum_rolling = $blockTickets->sum('rolling');
//            $sum_winnings = $blockTickets->sum('winnings');
//            $sum_bonus = $blockTickets->sum('bonus');
//
//            $first = $blockTickets->shift();
//
//            return [
//                'user_identify' => $first->user_identify,
//                'station' => $first->station,
//                'game_scope' => $first->game_scope,
//                'category' => $first->category,
//                'bet_type' => $first->bet_type,
//                'sum_raw_bet' => $sum_raw_bet,
//                'sum_valid_bet' => $sum_valid_bet,
//                'sum_rolling' => $sum_rolling,
//                'sum_winnings' => $sum_winnings,
//                'sum_bonus' => $sum_bonus,
//                'time_span_begin' => $timeStart,
//                'time_span_end' => $timeEnd,
//                'depth_0_identify' => isset($first->depth_0_identify) ? $first->depth_0_identify : '',
//                'depth_0_ratio' => isset($first->depth_0_ratio) ? $first->depth_0_ratio : 0.0,
//                'depth_1_identify' => isset($first->depth_1_identify) ? $first->depth_1_identify : '',
//                'depth_1_ratio' => isset($first->depth_1_ratio) ? $first->depth_1_ratio : 0.0,
//                'depth_2_identify' => isset($first->depth_2_identify) ? $first->depth_2_identify : '',
//                'depth_2_ratio' => isset($first->depth_2_ratio) ? $first->depth_2_ratio : 0.0,
//                'depth_3_identify' => isset($first->depth_3_identify) ? $first->depth_3_identify : '',
//                'depth_3_ratio' => isset($first->depth_3_ratio) ? $first->depth_3_ratio : 0.0,
//                'depth_4_identify' => isset($first->depth_4_identify) ? $first->depth_4_identify : '',
//                'depth_4_ratio' => isset($first->depth_4_ratio) ? $first->depth_4_ratio : 0.0,
//                'depth_5_identify' => isset($first->depth_5_identify) ? $first->depth_5_identify : '',
//                'depth_5_ratio' => isset($first->depth_5_ratio) ? $first->depth_5_ratio : 0.0,
//                'depth_6_identify' => isset($first->depth_6_identify) ? $first->depth_6_identify : '',
//                'depth_6_ratio' => isset($first->depth_6_ratio) ? $first->depth_6_ratio : 0.0,
//                'depth_7_identify' => isset($first->depth_7_identify) ? $first->depth_7_identify : '',
//                'depth_7_ratio' => isset($first->depth_7_ratio) ? $first->depth_7_ratio : 0.0,
//                'depth_8_identify' => isset($first->depth_8_identify) ? $first->depth_8_identify : '',
//                'depth_8_ratio' => isset($first->depth_8_ratio) ? $first->depth_8_ratio : 0.0,
//                'depth_9_identify' => isset($first->depth_9_identify) ? $first->depth_9_identify : '',
//                'depth_9_ratio' => isset($first->depth_9_ratio) ? $first->depth_9_ratio : 0.0,
//                'depth_10_identify' => isset($first->depth_10_identify) ? $first->depth_10_identify : '',
//                'depth_10_ratio' => isset($first->depth_10_ratio) ? $first->depth_10_ratio : 0.0,
//                'depth_11_identify' => isset($first->depth_11_identify) ? $first->depth_11_identify : '',
//                'depth_11_ratio' => isset($first->depth_11_ratio) ? $first->depth_11_ratio : 0.0,
//                'depth_12_identify' => isset($first->depth_12_identify) ? $first->depth_12_identify : '',
//                'depth_12_ratio' => isset($first->depth_12_ratio) ?  $first->depth_12_ratio : 0.0,
//            ];
//        });

        //---------------------------------------以下是使用SQL語法，速度比較慢------------------------------------------------------------------------

//        $united_tickets = DB::table('united_tickets')
//            ->where('invalid', false) // 未作廢，正常住單
//            ->where('compressed', false) // 未壓縮
//            ->where('payout_at', '<>', null) // 已開彩
//            ->where('station', '=', $station);
//
//        $this->united_ticket_ids = $united_tickets->get()->pluck('id');

        return DB::table('united_tickets')
            ->select(
                'user_identify', 'username', 'station', 'game_scope', 'bet_type', 'category',
                DB::raw('user_identify'),
                DB::raw('COUNT(id) AS ticket_count'),
                DB::raw('SUM(raw_bet) AS sum_raw_bet'),
                DB::raw('SUM(valid_bet) AS sum_valid_bet'),
                DB::raw('SUM(rolling) AS sum_rolling'),
                DB::raw('SUM(winnings) AS sum_winnings'),
                DB::raw('SUM(bonus) AS sum_bonus'),
                DB::raw('\'' . $timeStart . '\' AS time_span_begin'),
                DB::raw('\'' . $timeEnd . '\' AS time_span_end'),
                'depth_0_identify', 'depth_0_ratio',
                'depth_1_identify', 'depth_1_ratio',
                'depth_2_identify', 'depth_2_ratio',
                'depth_3_identify', 'depth_3_ratio',
                'depth_4_identify', 'depth_4_ratio',
                'depth_5_identify', 'depth_5_ratio',
                'depth_6_identify', 'depth_6_ratio',
                'depth_7_identify', 'depth_7_ratio',
                'depth_8_identify', 'depth_8_ratio',
                'depth_9_identify', 'depth_9_ratio',
                'depth_10_identify', 'depth_10_ratio'
//                'depth_11_identify', 'depth_11_ratio',
//                'depth_12_identify', 'depth_12_ratio'
            )
            ->groupBy(
                'user_identify', 'username', 'station', 'game_scope', 'bet_type', 'category',
                'depth_0_identify', 'depth_0_ratio',
                'depth_1_identify', 'depth_1_ratio',
                'depth_2_identify', 'depth_2_ratio',
                'depth_3_identify', 'depth_3_ratio',
                'depth_4_identify', 'depth_4_ratio',
                'depth_5_identify', 'depth_5_ratio',
                'depth_6_identify', 'depth_6_ratio',
                'depth_7_identify', 'depth_7_ratio',
                'depth_8_identify', 'depth_8_ratio',
                'depth_9_identify', 'depth_9_ratio',
                'depth_10_identify', 'depth_10_ratio'
//                'depth_11_identify', 'depth_11_ratio',
//                'depth_12_identify', 'depth_12_ratio'
            )
            ->where('invalid', false) // 未作廢，正常住單
            ->where('payout_at', '<>', null) // 已開彩
            ->where('payout_at', '>=', $timeStart)
            ->where('payout_at', '<', $timeEnd)
            ->where('station', '=', $station)
            ->get();
    }

    /**
     * @param $block
     * @param $timeStart
     * @return mixed
     */
    private function blockTransform($block, $timeStart)
    {
        return $block->transform(function ($item, $key) use ($timeStart) {
            $depth = '';

            for ($i = 0; $i < 13; $i++) {
                $depth_identify = 'depth_' . $i . '_identify';
                $depth = $depth . (empty($item->$depth_identify) ? '' : '-' . $item->$depth_identify);
            }

            $uniqueFieldsData = [
                $item->user_identify,
                $item->username,
                $item->station,
                $item->game_scope,
                $item->bet_type,
                $item->category,
                $item->time_span_begin,
                $item->time_span_end,
                $item->depth_0_identify,
                $item->depth_0_ratio,
                $item->depth_1_identify,
                $item->depth_1_ratio,
                $item->depth_2_identify,
                $item->depth_2_ratio,
                $item->depth_3_identify,
                $item->depth_3_ratio,
                $item->depth_4_identify,
                $item->depth_4_ratio,
                $item->depth_5_identify,
                $item->depth_5_ratio,
                $item->depth_6_identify,
                $item->depth_6_ratio,
                $item->depth_7_identify,
                $item->depth_7_ratio,
                $item->depth_8_identify,
                $item->depth_8_ratio,
                $item->depth_9_identify,
                $item->depth_9_ratio,
                $item->depth_10_identify,
                $item->depth_10_ratio,
//                $item->depth_11_identify,
//                $item->depth_11_ratio,
//                $item->depth_12_identify,
//                $item->depth_12_ratio,
            ];

            sort($uniqueFieldsData);
            $uniqueDataString = join('-', $uniqueFieldsData);

            $id = Uuid::uuid3(Uuid::NAMESPACE_DNS, $uniqueDataString . '@' . class_basename($this));

//            $item['id'] = $item['station'] . '-' . $item['game_scope'] . '-' . $item['bet_type'] . '-' . $item['user_identify'] . '-' . preg_replace('/[(\s)(:)]/i', '-', $timeStart) . $depth;

            $item->id = $id;
            return $item;
        });
    }
}